frontend/src/lib/StaticGraphParser.ts

```diff
@@ -76,72 +76,89 @@ export function _populateInfoFromTemplate(info: SelectedNodeInfo, template?: Tem
+  graph: dagre.graphlib.Graph,
+  rootTemplateId: string,
+  templates: Map<string, { nodeType: nodeType, template: Template }>,
+  alreadyVisited: Map<string, string>,
+  parentFullPath: string): void {
 
   const root = templates.get(rootTemplateId);
 
   if (root && root.nodeType === 'dag') {
+    // Mark that we have visited this DAG, and save the original qualified path to it.
+    alreadyVisited.set(rootTemplateId, parentFullPath || ('/' + rootTemplateId));
     const template = root.template;
 
     (template.dag.tasks || []).forEach((task) => {
 
-      const nodeId = (parent || '') + '/' + task.name;
+      const nodeId = parentFullPath + '/' + task.name;
 
       // If the user specifies an exit handler, then the compiler will wrap the entire Pipeline
       // within an additional DAG template prefixed with "exit-handler".
       // If this is the case, we simply treat it as the root of the graph and work from there
       if (task.name.startsWith('exit-handler')) {
-        buildDag(graph, task.template, templates);
-      } else {
-
-        // Parent here will be the task that pointed to this DAG template.
-        // We do not set an edge if the task has dependencies because in that case it makes sense
-        // for incoming edges to only be drawn from its dependencies, not the DAG node itself
-        if (parent && !task.dependencies) {
-          graph.setEdge(parent, nodeId);
-        }
+        buildDag(graph, task.template, templates, alreadyVisited, '');
+        return;
+      }
 
-        // This object contains information about the node that we display to the user when they
-        // click on a node in the graph
-        const info = new SelectedNodeInfo();
-        if (task.when) {
-          info.condition = task.when;
-        }
+      // If this task has already been visited, retrieve the qualified path name that was assigned
+      // to it, add an edge, and move on to the next task
+      if (alreadyVisited.has(task.name)) {
+        graph.setEdge(parentFullPath, alreadyVisited.get(task.name)!);
+        return;
+      }
 
-        // "Child" here is the template that this task points to. This template should either be a
-        // DAG, in which case we recurse, or a container which can be thought of as a leaf node
-        const child = templates.get(task.template);
-        if (child) {
-          if (child.nodeType === 'dag') {
-            buildDag(graph, task.template, templates, nodeId);
-          } else if (child.nodeType === 'container' ) {
-            _populateInfoFromTemplate(info, child.template);
-          } else {
-            logger.error(`Unknown nodetype: ${child.nodeType} on template: ${child.template}`);
-          }
-        }
+      // Parent here will be the task that pointed to this DAG template.
+      // We do not set an edge if the task has dependencies because in that case it makes sense
+      // for incoming edges to only be drawn from its dependencies, not the DAG node itself
```

Comment:
Can you clarify this comment a little? Why does having or not having dependencies affect having an edge to its parent node?

Reply:
This might be a bit wordy, see if you think it makes sense:
```
// Parent here will be the task that pointed to this DAG template.
// Within a DAG template, tasks can have dependencies on one another, and long chains of
// dependencies can be present within a single DAG. In these cases, we choose not to draw an
// edge from the DAG node itself to these tasks with dependencies because such an edge would
// not be meaningful to the user. For example, consider a DAG A with two tasks, B and C, where
// task C depends on the output of task B. C is a task of A, but it's much more important that
// C depends on B, so to avoid cluttering the graph, we simply omit the edge between A and C:
//      A                  A   
//    /   \    becomes    /  
//   B <-- C             B 
//                      /
//                     C

Reply:
Are we sure we won't ever be removing useful information from the graph by doing so?
Separately, the comment is much more clear now, thanks.

Reply:
I don't know for certain, but I haven't been able to think of a case where it'd be preferable to see those extra edges, and removing them makes the execution flow of the pipeline a lot clearer (IMO)

Reply:
SGTM then, thanks.
/lgtm